### ISSUE TITLE
Specify network adapter type for primary NIC and also additional NICs

### ISSUE THREAD

**Issue by mingsheng36** [DimensionDataResearch/terraform-provider-ddcloud#52]
Apologize for spamming the issues section. 

As you know our NIC adapter will always default to E1000 for almost all the images. I have a client that requires the use of vmxnet3 for better network performances. Looking at the codes, seems like this request will involve adding codes at the GO compute..

**Comment by wninobla** [comment]
Hi Ming,

I think by default the template image is going to use whatever the default is which I believe the VMNET3 NIC it is.  But, you bring up a valid point for people who need the older driver.

Adam,

Any way we can add this option into the server deployment resource?  Whether we deploy it initially or with the NIC resource add we'll need a way to specify that for legacy (or poorly coded ones).  Thoughts?

**Comment by mingsheng36** [comment]
[https://docs.mcp-services.net/display/CCD/What+Operating+Systems+are+Currently+Supported](url)

For the Linux OSes, all are using E1000 as their default.

Thanks William for the suggestion. To me, this is an important factor as if there is requirement for vmxnet3 (or any other adapter types other than the default), then terraform can't really be used for automated provisioning.

**Comment by wninobla** [comment]
While the push should be to use the latest NIC I do agree there will undoubtedly be edge cases where we will need this.

**Comment by tintoy** [comment]
Hi - sure, that should be pretty easy to do. I'll have a look into this today.

- [x] `VirtualMachineNetworkAdapter.AdapterType`
- [x] `ddcloud_server.primary_adapter_type`
  - [x] Implementation
  - [x] Documentation
- [x] `ddcloud_server_nic.adapter_type`
  - [x] Implementation
  - [x] Documentation

**Comment by wninobla** [comment]
Thanks Adam!

I’ve been sick so I need to test a few things now and also add 32-bit testing to the fold as well.  If I can get this GoCD crap to work I’d be set =)  Once I see the commit I’ll give it a whirl and update my docs.  Thanks!

**Comment by tintoy** [comment]
Hmm. Actually, the CloudControl API doesn't seem to let you specify the NIC type anywhere that I can see :-/

I'll ask the CaaS UI guys how they do it.

**Comment by wninobla** [comment]
?Yeah - I think it defaults to auto detact in the UI so I know theres "something" though it may not be API accessible.  Good luck! =)

**Comment by tintoy** [comment]
It looks like the ability to specify NIC type is only present in the 2.3 API but we use 2.2 to deploy servers. I'll need to see if this breaks compatibility with anything before we can proceed.

**Comment by tintoy** [comment]
Ok - this should work as of [v1.1.5](https://github.com/DimensionDataResearch/dd-cloud-compute-terraform/releases/tag/v1.1.5).

* `ddcloud_server` now has property `primary_adapter_type`
* `ddcloud_server_nic` now has property `adapter_type`

Both can be set to either "E1000" or "VMXNET3".

Due to limitations in the CloudControl API, we can only _set_ the adapter type; it is not visible when retrieving the server details (so we never refresh this property but since it's read-only after creation this should not be a problem).

@mingsheng36 and / or @wninobla can you give this a try and let me know if it works for you?

**Comment by mingsheng36** [comment]
![image](https://cloud.githubusercontent.com/assets/5947212/20506102/89737d6a-b08c-11e6-90f3-79557a586862.png)
![image](https://cloud.githubusercontent.com/assets/5947212/20506119/a3adb024-b08c-11e6-86a1-018e7e224757.png)

It is working now! Wonderful.

Thanks @wninobla for pushing it and @tintoy for the quick response and fix! :)

**Comment by wninobla** [comment]
You sir are quick. =)  I should probably stop playing with GoCD and get to testing myself. =)

**Comment by mingsheng36** [comment]
While trying to add additional nic through "server_nic.tf", it does not deploy it as VMXNET3.

resource "ddcloud_server_nic" "test-server-2-nic" {
  server               = "${ddcloud_server.test-server-2.id}"
  private_ipv4         = "172.16.0.100"
  adapter_type 		   = "VMXNET3"
  vlan                 = "b6dfce96-b98c-4eea-9968-854a65acd03d"
  depends_on           = ["ddcloud_server.test-server-2"]
}

And also "allow_server_reboot" in provider settings does not seem to work while trying to update from E1000 -> VMXNET3 for additional NICs.

provider "ddcloud" {
  "username"           = "YOUR USERNAME"
  "password"           = "YOUR PASSWORD"
  "region"             = "AP"
  "allow_server_reboot" = "TRUE"
}

**Comment by mingsheng36** [comment]
When running "plan":
![image](https://cloud.githubusercontent.com/assets/5947212/20513240/a0cdd87e-b0bf-11e6-837e-21a462c8fde9.png)

Actual code:
![image](https://cloud.githubusercontent.com/assets/5947212/20513247/ab759f00-b0bf-11e6-8244-6cc6974aa0b9.png)

**Comment by tintoy** [comment]
Regarding `allow_server_reboot`, would you mind raising a separate issue for that with a description of the behaviour you're seeing? It's easier to keep track of if each issue in GitHub relates to a single thing :)

**Comment by mingsheng36** [comment]
Sure thing. Let me test it out.

**Comment by tintoy** [comment]
@mingsheng36 I think I've found a bug there (`allow_server_reboot`). Will be fixed in v1.1.6.

**Comment by tintoy** [comment]
@mingsheng36 - also, the problem with `adapter_type` in `ddcloud_server_nic` should be fixed in v1.1.6, too.

**Comment by tintoy** [comment]
Ok, give [v1.1.6](https://github.com/DimensionDataResearch/dd-cloud-compute-terraform/releases/tag/v1.1.6) a try.

**Comment by mingsheng36** [comment]
Will test both of them tmr. 👍

**Comment by wninobla** [comment]
I am finalizing the GoCD implementation (I have to figure out the variable fields to pass from GoCD but once I do I’ll have this running as a continuous cycle for Windows and Linux testing. =)

**Comment by tintoy** [comment]
Awesome :)

BTW check out our Travis build file if you need any pointers.

**Comment by mingsheng36** [comment]
Hey Adam,

Sorry for the late reply. 

Since "allow_server_reboot" has been fixed, I shall not open another issue for that.

Seems like when we add the additional NIC and specify VMXNET3, it is still adding it as E1000.

.tf:
![image](https://cloud.githubusercontent.com/assets/5947212/20611774/a3594a96-b2eb-11e6-907f-9c0c522eeee7.png)

os:
![image](https://cloud.githubusercontent.com/assets/5947212/20611784/c86262be-b2eb-11e6-9110-d13e155735d7.png)

**Comment by tintoy** [comment]
What version of the provider are you now using? You can run `terraform-provider-ddcloud --version` to print that out.

Can you send me the full configuration you’re working with? The one I used was working for me so I’d like to have a play around with yours to see if I can replicate the issue in a controlled environment.

**Comment by mingsheng36** [comment]
Their latest version:
![image](https://cloud.githubusercontent.com/assets/5947212/20611884/3f7b3514-b2ed-11e6-9b61-4c3c41f41a2b.png)

For terraform-provider-ddcloud.exe (64-bit), I am using v1.1.7 (latest)

**Comment by tintoy** [comment]
Sorry, I meant can you run  `terraform-provider-ddcloud —version` not `terraform version` (it'll include the exact commit that it was built from).

**Comment by mingsheng36** [comment]
My bad.

For terraform-provider-ddcloud.exe (64-bit), I am using v1.1.7 (latest)

![image](https://cloud.githubusercontent.com/assets/5947212/20611936/4660305e-b2ee-11e6-9f40-904b89efdf22.png)

**Comment by tintoy** [comment]
Ok - send me your configuration and I’ll have a look.

**Comment by mingsheng36** [comment]
.tf
![image](https://cloud.githubusercontent.com/assets/5947212/20612056/a0424006-b2f0-11e6-9d0c-c058bc9e4f13.png)

terraform plan:
![image](https://cloud.githubusercontent.com/assets/5947212/20612069/c3644282-b2f0-11e6-9bfa-5a5a17a7c1fa.png)

terraform apply:
![image](https://cloud.githubusercontent.com/assets/5947212/20612176/27f9f6f0-b2f2-11e6-98f7-c6465a83b68a.png)

terraform state:
![image](https://cloud.githubusercontent.com/assets/5947212/20612183/42743432-b2f2-11e6-85d0-9d9e935d8077.png)

OS check: 
![image](https://cloud.githubusercontent.com/assets/5947212/20612253/dfc1c862-b2f2-11e6-99aa-6561a573a133.png)

**Comment by tintoy** [comment]
Perhaps text rather than screenshots? I can't run a screenshot :)

**Comment by tintoy** [comment]
BTW, if possible could you send the entire configuration rather than just a subset? The easier it is for me to get it running the sooner I can reproduce the problem :)

**Comment by mingsheng36** [comment]
sure.

**Comment by mingsheng36** [comment]
```
resource "ddcloud_server_nic" "docker-server" {
  server               = "1990631e-5545-43a0-bb99-d741b0d48f00"
  private_ipv4         = "10.0.0.100"
  adapter_type		   = "VMXNET3"
}

resource "ddcloud_server_nic" "pomelo-server" {
  server               = "06a08423-7e06-4bd7-99a8-86b1eb5d0606"
  private_ipv4         = "172.16.0.100"
  adapter_type		   = "VMXNET3"
}
```

**Comment by tintoy** [comment]
Sorry to be a pain, but would you mind including the *full* configuration (i.e. all resources not just those 2)? I want to make sure I’m replicating your configuration exactly.

**Comment by tintoy** [comment]
Feel free to email it to me directly if you're worried it has something semi-sensitive in it.

**Comment by mingsheng36** [comment]
Sure. For this scenario as you can see i have manually added the Server ID which means it is an already deployed server which is not using terraform. Hence the "server_nic.tf" was the only thing that I have included.

**Comment by tintoy** [comment]
I see. Ok, I'll give that a try.

**Comment by tintoy** [comment]
Ok, I figured out why it works for the primary NIC but not additional ones.

Turns out that the v2.3 API (which we use to deploy servers) respects the adapter type, but the v2.2 API (which we use to add NICs to servers) does not. I modified our CloudControl client to use the v2.3 API for adding NICs and that did the trick.

Unfortunately this now means that you can specify the adapter's IPv4 address or VLAN, but not both (not that it matters much - specifying one implies the other).

**Comment by tintoy** [comment]
Ok @mingsheng36 give [v1.1.8](https://github.com/DimensionDataResearch/dd-cloud-compute-terraform/releases/tag/v1.1.8) a try. Be aware that now you can only specify `vlan` or `primary_ipv4` but not both (specifying `primary_ipv4` is preferred).

**Comment by mingsheng36** [comment]
Adam, just a minor request. Do you mind including the version number on the releases file names? It would help alot with identifying the files i have downloaded

**Comment by tintoy** [comment]
Hi.

Sorry, that’s actually a bit difficult given the release process at present. I’ll see if I can improve that in the future but for now at least you can run “terraform-provider-ddcloud --version” which will tell you exactly what version it s.

**Comment by tintoy** [comment]
BTW, would you mind creating separate issues for changes of topic? It makes it a lot easier to track progress :)

I've moved this discussion to #55.

**Comment by tintoy** [comment]
@mingsheng36 - FYI, I've done a quick-fix so the packages are generated with more useful names. I'll revisit that later to make it less dodgy but it'll do for now.

**Comment by mingsheng36** [comment]
Solved thanks!
